File: python/mirage/core/mongodb/readdir.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.mongodb import MongoDBAccessor
from mirage.cache.index import IndexCacheStore, IndexEntry
from mirage.core.mongodb._client import (database_exists, entity_exists,
                                         list_collections, list_databases)
from mirage.core.mongodb.scope import detect_scope
from mirage.core.mongodb.types import (KIND_TO_DIR, KIND_TO_RESOURCE_TYPE,
                                       RESOURCE_TYPE_DATABASE, EntityKind,
                                       ScopeLevel)
from mirage.types import PathSpec


def is_dir_name(child: str) -> bool:
    # Entries are recognized by extension, so classification never needs the
    # stat fallback.
    name = child.rsplit("/", 1)[-1]
    return not (name.endswith(".json") or name.endswith(".jsonl"))


async def readdir(
    accessor: MongoDBAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> list[str]:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    prefix = path.prefix or ""
    scope = detect_scope(path)
    virtual_key = (prefix + scope.resource_path).rstrip("/") or "/"

    if scope.level == ScopeLevel.ROOT:
        return await _list_root(accessor, virtual_key, index, prefix)

    if scope.level == ScopeLevel.DATABASE:
        if not await database_exists(accessor.client, accessor.config,
                                     scope.database, accessor):
            raise FileNotFoundError(path.original)
        base = f"{prefix}/{scope.database}"
        return [
            f"{base}/database.json",
            f"{base}/collections",
            f"{base}/views",
        ]

    if scope.level == ScopeLevel.KIND_DIR:
        if not await database_exists(accessor.client, accessor.config,
                                     scope.database, accessor):
            raise FileNotFoundError(path.original)
        return await _list_kind_dir(accessor, scope.database, scope.kind,
                                    virtual_key, index, prefix)

    if scope.level == ScopeLevel.ENTITY:
        if not await entity_exists(accessor.client, accessor.config,
                                   scope.database, scope.name, scope.kind,
                                   accessor):
            raise FileNotFoundError(path.original)
        base = (f"{prefix}/{scope.database}/"
                f"{KIND_TO_DIR[scope.kind]}/{scope.name}")
        return [
            f"{base}/schema.json",
            f"{base}/documents.jsonl",
        ]

    raise FileNotFoundError(path.original)


async def _list_root(
    accessor: MongoDBAccessor,
    virtual_key: str,
    index: IndexCacheStore | None,
    prefix: str,
) -> list[str]:
    if index is not None:
        listing = await index.list_dir(virtual_key)
        if listing.entries is not None:
            return listing.entries
    dbs = await list_databases(accessor.client, accessor.config)
    entries: list[tuple[str, IndexEntry]] = []
    names: list[str] = []
    for db_name in dbs:
        entry = IndexEntry(
            id=db_name,
            name=db_name,
            resource_type=RESOURCE_TYPE_DATABASE,
            vfs_name=db_name,
        )
        entries.append((db_name, entry))
        names.append(f"{prefix}/{db_name}")
    if index is not None:
        await index.set_dir(virtual_key, entries)
    return names


async def _list_kind_dir(
    accessor: MongoDBAccessor,
    database: str,
    kind: EntityKind,
    virtual_key: str,
    index: IndexCacheStore | None,
    prefix: str,
) -> list[str]:
    if index is not None:
        listing = await index.list_dir(virtual_key)
        if listing.entries is not None:
            return listing.entries
    names = await list_collections(accessor.client, database, kind=kind)
    base = f"{prefix}/{database}/{KIND_TO_DIR[kind]}"
    entries: list[tuple[str, IndexEntry]] = []
    out: list[str] = []
    for name in names:
        entry = IndexEntry(
            id=name,
            name=name,
            resource_type=KIND_TO_RESOURCE_TYPE[kind],
            vfs_name=name,
        )
        entries.append((name, entry))
        out.append(f"{base}/{name}")
    if index is not None:
        await index.set_dir(virtual_key, entries)
    return out

File: typescript/packages/core/src/core/mongodb/readdir.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { MongoDBAccessor } from '../../accessor/mongodb.ts'
import { IndexEntry } from '../../cache/index/config.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { PathSpec } from '../../types.ts'
import { databaseExists, entityExists, listCollections, listDatabases } from './_client.ts'
import { detectScope } from './scope.ts'
import type { EntityKind } from './types.ts'
import { KIND_TO_DIR, KIND_TO_RESOURCE_TYPE, RESOURCE_TYPE_DATABASE, ScopeLevel } from './types.ts'
import { rstripSlash } from '../../utils/slash.ts'

function notFound(p: string): Error {
  const err = new Error(p) as Error & { code?: string }
  err.code = 'ENOENT'
  return err
}

export async function readdir(
  accessor: MongoDBAccessor,
  path: PathSpec | string,
  index?: IndexCacheStore,
): Promise<string[]> {
  const spec = typeof path === 'string' ? PathSpec.fromStrPath(path) : path
  const prefix = spec.prefix
  const scope = detectScope(spec)
  const virtualKey = rstripSlash(`${prefix}${scope.resourcePath}`) || '/'

  if (scope.level === ScopeLevel.ROOT) {
    return listRoot(accessor, virtualKey, index, prefix)
  }

  if (scope.level === ScopeLevel.DATABASE && scope.database !== null) {
    if (!(await databaseExists(accessor, scope.database))) throw notFound(spec.original)
    const base = `${prefix}/${scope.database}`
    return [`${base}/database.json`, `${base}/collections`, `${base}/views`]
  }

  if (scope.level === ScopeLevel.KIND_DIR && scope.database !== null && scope.kind !== null) {
    if (!(await databaseExists(accessor, scope.database))) throw notFound(spec.original)
    return listKindDir(accessor, scope.database, scope.kind, virtualKey, index, prefix)
  }

  if (
    scope.level === ScopeLevel.ENTITY &&
    scope.database !== null &&
    scope.kind !== null &&
    scope.name !== null
  ) {
    if (!(await entityExists(accessor, scope.database, scope.name, scope.kind))) {
      throw notFound(spec.original)
    }
    const base = `${prefix}/${scope.database}/${KIND_TO_DIR[scope.kind]}/${scope.name}`
    return [`${base}/schema.json`, `${base}/documents.jsonl`]
  }

  throw notFound(spec.original)
}

async function listRoot(
  accessor: MongoDBAccessor,
  virtualKey: string,
  index: IndexCacheStore | undefined,
  prefix: string,
): Promise<string[]> {
  if (index !== undefined) {
    const cached = await index.listDir(virtualKey)
    if (cached.entries !== null && cached.entries !== undefined) return cached.entries
  }
  const dbs = await listDatabases(accessor)
  const entries: [string, IndexEntry][] = []
  const names: string[] = []
  for (const db of dbs) {
    entries.push([
      db,
      new IndexEntry({
        id: db,
        name: db,
        resourceType: RESOURCE_TYPE_DATABASE,
        vfsName: db,
      }),
    ])
    names.push(`${prefix}/${db}`)
  }
  if (index !== undefined) await index.setDir(virtualKey, entries)
  return names
}

async function listKindDir(
  accessor: MongoDBAccessor,
  database: string,
  kind: EntityKind,
  virtualKey: string,
  index: IndexCacheStore | undefined,
  prefix: string,
): Promise<string[]> {
  if (index !== undefined) {
    const cached = await index.listDir(virtualKey)
    if (cached.entries !== null && cached.entries !== undefined) return cached.entries
  }
  const names = await listCollections(accessor, database, kind)
  const base = `${prefix}/${database}/${KIND_TO_DIR[kind]}`
  const entries: [string, IndexEntry][] = []
  const out: string[] = []
  for (const name of names) {
    entries.push([
      name,
      new IndexEntry({
        id: name,
        name,
        resourceType: KIND_TO_RESOURCE_TYPE[kind],
        vfsName: name,
      }),
    ])
    out.push(`${base}/${name}`)
  }
  if (index !== undefined) await index.setDir(virtualKey, entries)
  return out
}
